### PR TITLE
Dont crash immediately on big opencl errors, closes #5

### LINKED ISSUE #5: OpenCL calculation error
Having played 1912 games, 607 had an error > 5% in the OpenCL calculation. The average OpenCL error was 26.5%. Do others experience this also?

`grep 'error=' training.out | cut -d' ' -f11 | cut -d'=' -f2 | cut -d'%' -f1 > errors.txt`

``` octave
>> load errors.txt
>> mean(errors)
ans =  26.452
>> std(errors)
ans =  56.250
>> min(errors)
ans =  5.0110
>> max(errors)
ans =  632.42
```

#### System configuration:
Ubuntu 16.04
Intel(R) Core(TM) i7-6800K CPU @ 3.40GHz
Platform version: OpenCL 1.2 CUDA 9.1.98
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     0
Device name:   GeForce GTX 1080 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 387.34
Device speed:  1657 MHz
Device cores:  28 CU
Device score:  1112
Device ID:     1
Device name:   GeForce GTX 1080 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 387.34
Device speed:  1657 MHz
Device cores:  28 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1080 Ti
with OpenCL 1.2 capability
Wavefront/Warp size: 32
Max workgroup size: 1024
Max workgroup dimensions: 1024 1024 64

### PR DESCRIPTION
**Pull request by Error323**
Please verify for correctness.

### CODE REVIEW

File: src/config.h [hunk 1]
```diff
@@ -33,7 +33,8 @@
 //#define USE_MKL
 #define USE_OPENCL
 #define USE_OPENCL_SELFCHECK
-#define SELFCHECK_PROBABILITY 2000
+static constexpr int SELFCHECK_PROBABILITY = 2000;
+static constexpr int SELFCHECK_MIN_EXPANSIONS = 2'000'000;
```
Comment (Error323) [inline src/config.h]: Not sure about the value of this number yet. Maybe this is a bit high?

**Comment by Error323** [comment]
Oops, it's still bugged hold on...

File: src/Network.cpp [hunk 2]
```diff
@@ -869,17 +869,28 @@ T relative_difference(T a, T b) {
...
-                   "played simultaneously.\n");
-            throw std::runtime_error("OpenCL self-check mismatch.");
+        if (err > relative_error) {
+            if (num_expansions < min_correct_expansions) {
+                printf("Update your GPU drivers or reduce the amount of games "
+                           "played simultaneously.\n");
+                throw std::runtime_error("OpenCL self-check mismatch.");
+            }
+            else {
+                num_expansions -= min_correct_expansions;
```
Comment (glinscott) [inline src/Network.cpp]: Hm..  Can we comment the intent here?  I don't quite understand what the check is doing.

File: src/Network.cpp [hunk 3]
```diff
@@ -869,17 +869,28 @@ T relative_difference(T a, T b) {
...
+    num_expansions++;
+
     // We accept an error up to 5%, but output values
     // smaller than 1/1000th are "rounded up" for the comparison.
     constexpr float relative_error = 5e-2f;
     for (auto idx = size_t{0}; idx < data.size(); ++idx) {
         auto err = relative_difference(data[idx], ref[idx]);
-        if (err > relative_error) {
-            printf("Error in OpenCL calculation: expected %f got %f "
+        myprintf("Error in OpenCL calculation: expected %f got %f "
```
Comment (glinscott) [inline src/Network.cpp]: I think this is going to print even if there is no error?

File: src/Network.cpp [hunk 4]
```diff
@@ -869,17 +869,28 @@ T relative_difference(T a, T b) {
 
 void compare_net_outputs(std::vector<float>& data,
                          std::vector<float>& ref) {
+    // Divide by an additional 2 as we're comparing 2 heads at a time
+    constexpr int64 min_correct_expansions = SELFCHECK_MIN_EXPANSIONS / SELFCHECK_PROBABILITY / 2;
+    static_assert(min_correct_expansions > 0, "Increase minimal nof expansions");
+    static int64 num_expansions{min_correct_expansions};
```
Comment (glinscott) [inline src/Network.cpp]: This could be problematic for multi-threaded evaluation.

File: src/Network.cpp (hunk 2 above)
Comment (Error323) [reply src/Network.cpp]: The idea is to allow an OpenCL error > 5% every 2'000'000 correct expansions. As the nof expansions increases between errors > 5%, we'll allow more errors to occur before crashing. As if it builds up credit.

File: src/Network.cpp (hunk 3 above)
Comment (Error323) [reply src/Network.cpp]: The idea was to gather statistics on the error size.

File: src/Network.cpp (hunk 4 above)
Comment (Error323) [reply src/Network.cpp]: Right! I guess this should be atomic.

Comment (Error323) [reply src/Network.cpp]: Right! I guess this should be atomic.

**Comment by Error323** [comment]
Are you in agreement with this now @glinscott ? What's your take on it @killerducky ?

**Comment by killerducky** [comment]
I think it's ok for short-term. Long-term I hope we can debug the root cause of the issue. So far it seems to only happens on some systems? I tried running 4 clients on my machine for a short time and didn't see it, does it take a long time to hit it? A small number of LZGo users have reported similar issues, and they workaround it by just running fewer clients in parallel. The majority of LZGo users never see the problem. 

I propose we take this pull for now, and next step work on getting client games to be reproducible (run with 1 thread), record all command line args including random seed to lzchess in the pgn or as metadata sent back to the server. Maybe also add logging that an OpenCL miscompare happened in this game.

This way as users get the error we can try to reproduce and debug. I'm afraid it will end up being system load dependent though. :(

**Comment by Error323** [comment]
Agreed. I'll fiddle more with drivers on my end. It happened every hour or so with 2 instances, one on each GPU.

File: src/Network.cpp (hunk 2 above)
Comment (glinscott) [reply src/Network.cpp]: Makes sense, thanks for adding the comment!

**Comment by glinscott** [comment]
@Error323 looks good to me, feel free to merge any time.

**Comment by Error323** [comment]
When linking with the specific cuda version of OpenCL instead of the default one in Ubuntu I replayed a deterministic run between `lc-64x6` and `lc-8x0`. Every game had exactly the same result this time.

```
$ pgn-extract -D results.pgn > out.pgn
Processing results.pgn
lc_64x6 - lc_8x0 ? ? 2018.03.02 
lc_8x0 - lc_64x6 ? ? 2018.03.02 
2 games matched out of 20.
```

Unfortunately when running multiple clients again for longer I did still experience a > 5% error occurring. So this PR has to remain I'm afraid.

Running latest driver as of this moment:

```
Detected 1 OpenCL platforms.
Platform version: OpenCL 1.2 CUDA 9.1.84
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     0
Device name:   GeForce GTX 1080 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 390.30
Device speed:  1657 MHz
Device cores:  28 CU
Device score:  1112
Device ID:     1
Device name:   GeForce GTX 1080 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 390.30
Device speed:  1657 MHz
Device cores:  28 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1080 Ti
with OpenCL 1.2 capability.
```

**Pull request merged**